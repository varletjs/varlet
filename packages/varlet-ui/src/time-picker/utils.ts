import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
