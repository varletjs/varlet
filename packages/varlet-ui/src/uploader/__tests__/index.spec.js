import example from '../example'
import Uploader from '..'
import VarUploader from '../Uploader'
import ImagePreview from '../../image-preview'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, mockFileReader, mockStubs } from '../../utils/jest'

const createEvent = (filename) => {
  return {
    target: {
      files: [new File([], filename)],
    },
  }
}

test('test uploader example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test uploader plugin', () => {
  const app = createApp({}).use(Uploader)
  expect(app.component(Uploader.name)).toBeTruthy()
})

test('test uploader onAfterRead', async () => {
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      onAfterRead,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test uploader onBeforeRead', async () => {
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      onAfterRead,
      onBeforeRead: (file) => file.name.endsWith('jpg'),
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(0)

  await wrapper.vm.handleChange(createEvent('cat.jpg'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test uploader preview', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const { mockRestore: mockRestoreStubs } = mockStubs()
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarUploader, {
    props: {
      modelValue: [],
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.jpg'))
  await delay(16)
  await wrapper.find('.var-uploader__file').trigger('click')
  await delay(100)

  expect(document.querySelector('.var-popup').style.display).toBe('')
  ImagePreview.close()
  await delay(300)

  mockRestoreStubs()
  wrapper.unmount()
  mockRestore()
})

test('test uploader onOversize', async () => {
  const onOversize = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      maxsize: -1,
      onOversize,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onOversize).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test uploader onRemove', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))
  const onRemove = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      modelValue: [],
      onRemove,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onUpdateModelValue).toHaveBeenCalledTimes(1)

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onUpdateModelValue).toHaveBeenCalledTimes(2)
  expect(onRemove).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.modelValue).toStrictEqual([])

  wrapper.unmount()
  mockRestore()
})

test('test uploader onBeforeRemove', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))
  const onBeforeRemove = jest.fn(() => false)
  const onRemove = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      modelValue: [],
      onBeforeRemove,
      onRemove,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onUpdateModelValue).toHaveBeenCalledTimes(1)

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onBeforeRemove).toHaveBeenCalledTimes(1)
  expect(onUpdateModelValue).toHaveBeenCalledTimes(1)
  expect(onRemove).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.modelValue).toHaveLength(1)

  wrapper.unmount()
  mockRestore()
})

test('test uploader validation', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarUploader, {
    props: {
      modelValue: [],
      rules: [(v) => v.length >= 1 || '您至少上传一个'],
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  wrapper.vm.validate()
  await delay(16)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-form-details__message').text()).toBe('您至少上传一个')

  await wrapper.vm.handleChange(createEvent('cat.png'))
  await delay(16)
  expect(onUpdateModelValue).toHaveBeenCalledTimes(1)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()

  wrapper.vm.reset()
  await delay(16)
  expect(wrapper.vm.modelValue).toStrictEqual([])

  wrapper.unmount()
  mockRestore()
})

test('test uploader disabled', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))
  const onRemove = jest.fn()
  const onAfterRead = jest.fn()

  const wrapper = mount(VarUploader, {
    props: {
      disabled: false,
      modelValue: [],
      onAfterRead,
      onRemove,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
  await wrapper.vm.handleChange(createEvent('cat.png'))
  expect(onAfterRead).toHaveBeenCalledTimes(1)

  await wrapper.setProps({ disabled: true })
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-uploader__action-input[disabled]').exists()).toBeTruthy()

  await wrapper.find('.var-uploader__file-close').trigger('click')
  expect(onRemove).toHaveBeenCalledTimes(0)

  wrapper.unmount()
  mockRestore()
})

test('test uploader length over maxlength in multiple mode', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarUploader, {
    props: {
      maxlength: 1,
      modelValue: [],
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  const event = {
    target: {
      files: [new File([], 'cat.png'), new File([], 'dog.png')],
    },
  }

  await wrapper.vm.handleChange(event)
  expect(wrapper.vm.modelValue.length).toBe(1)

  wrapper.unmount()
  mockRestore()
})

test('test uploader hide file list', async () => {
  const { mockRestore } = mockFileReader('data:image/png;base64,')
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarUploader, {
    props: {
      hideList: true,
      modelValue: [],
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  const event = {
    target: {
      files: [new File([], 'cat.png'), new File([], 'dog.png')],
    },
  }

  await wrapper.vm.handleChange(event)
  expect(wrapper.vm.renderFileList.length).toBe(0)

  wrapper.unmount()
  mockRestore()
})
