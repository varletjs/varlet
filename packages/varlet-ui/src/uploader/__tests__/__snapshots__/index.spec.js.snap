// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test uploader disabled 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader disabled 2`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">cat.png</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
      <!--v-if-->
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2 var-uploader--disabled\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\" disabled=\\"\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">文件预览</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://www.runoob.com/try/demo_source/mov_bbb.mp4</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cover.jpg\\">
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">上传状态</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2 var-uploader--loading\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator var-uploader--success\\"></div>
    </div>
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator var-uploader--error\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">文件数量限制</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\"></div>
      <div class=\\"var-form-details__length\\">0 / 1</div>
    </div>
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">文件大小限制</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">上传预处理</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">禁用</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2 var-uploader--disabled\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\" disabled=\\"\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">只读</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\" disabled=\\"\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">删除前处理</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">自定义上传样式</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2\\">
        <!--v-if-->
        <div class=\\"var-button__content\\">上传</div>
      </button></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">字段校验</div>
<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">https://varlet-varletjs.vercel.app/cat.jpg</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div><img class=\\"var-uploader__file-cover\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
      <div class=\\"var-uploader__file-indicator var-uploader--error\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>
<div class=\\"app-type\\">自定义渲染文件列表</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 12px 4px 0px;\\"><img class=\\"custom-uploader-file\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\"></div>
  <div style=\\"margin: 4px 12px 4px 0px;\\"><img class=\\"custom-uploader-file\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\"></div>
  <div style=\\"margin: 4px 12px 4px 0px;\\"><img class=\\"custom-uploader-file\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\"></div>
  <div style=\\"margin: 4px 0px;\\">
    <div class=\\"var-uploader var--box\\">
      <div class=\\"var-uploader__file-list\\">
        <div class=\\"\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2 var-button--round\\">
            <!--v-if-->
            <div class=\\"var-button__content\\"><i class=\\"var-icon var-icon--set var-icon-upload\\" style=\\"transition: transform 0ms; font-size: 28px;\\"></i></div>
          </button></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
      <transition-stub>
        <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
          <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
          <transition-stub>
            <!---->
          </transition-stub>
        </div>
      </transition-stub>
    </div>
  </div>
</div>
<div class=\\"space\\"></div>"
`;

exports[`test uploader hide file list 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader length over maxlength in multiple mode 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\"></div>
      <div class=\\"var-form-details__length\\">0 / 1</div>
    </div>
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader validation 1`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">您至少上传一个</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;

exports[`test uploader validation 2`] = `
"<div class=\\"var-uploader var--box\\">
  <div class=\\"var-uploader__file-list\\">
    <div class=\\"var-uploader__file var-elevation--2\\">
      <div class=\\"var-uploader__file-name\\">cat.png</div>
      <div class=\\"var-uploader__file-close\\"><i class=\\"var-icon var-icon--set var-icon-delete var-uploader__file-close-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
      <!--v-if-->
      <div class=\\"var-uploader__file-indicator\\"></div>
    </div>
    <div class=\\"var-uploader__action var-elevation--2\\"><input class=\\"var-uploader__action-input\\" type=\\"file\\" accept=\\"image/*\\"><i class=\\"var-icon var-icon--set var-icon-plus var-uploader__action-icon\\" style=\\"transition: transform 0ms;\\" var-uploader-cover=\\"\\"></i></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <transition-stub>
    <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
      <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
      <transition-stub>
        <!---->
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;
